[HttpFoundation] Make sessions secure and lazy

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | not yet
| Fixed tickets | #6388, #6036, #12375, #12325
| License       | MIT
| Doc PR        | -

The `SessionUpdateTimestampHandlerInterface` (new to PHP 7.0) is mostly undocumented, and just not implemented anywhere. Yet, it's required to implement session fixation preventions and lazy write in userland session handlers (there is https://wiki.php.net/rfc/session-read_only-lazy_write which describes the behavior.)

By implementing it, we would make Symfony session handling much better and stronger. Meanwhile, doing some cookie headers management, this also gives the opportunity to fix the "don't start if session is only read issue".

So, here we are for the general idea. Now needs more (and green) tests, and review of course.

nicolas-grekas committed:
[HttpFoundation] Make sessions secure and lazy

eXtreme commented:
wow, there's literally nothing on google about `SessionUpdateTimestampHandlerInterface`, yet it exists for so long ..

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php:
```diff
@@ -517,7 +542,9 @@ private function doRead($sessionId)
-            if (self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+            if (\PHP_VERSION_ID < 70000 && self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+                // Before PHP 7.0, session fixation was possible so locking could be needed even when creating a session.
+                // Starting with 7.0, secure random ids are generated so not concurrency is possible, thus this code path can be removed.
```
@Tobion do you agree with this statement?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml:
```diff
@@ -48,11 +48,17 @@
-            <argument>%session.save_path%</argument>
+        <service id="session.handler.native_file" class="Symfony\Component\HttpFoundation\Session\Storage\Handler\SessionHandlerProxy">
+            <argument type="service">
+                <service class="Symfony\Component\HttpFoundation\Session\Storage\Handler\NativeFileSessionHandler">
```
for some obscure reason, the native `SessionHandler` doesn't implement `SessionUpdateTimestampHandlerInterface`
we have to proxy it to add strict mode and lazy writes...

nicolas-grekas commented:
PR is green and ready. This is a significant improvement over the current session handlers, thanks to this new PHP 7.0 interface.

This adds two main classes:
- `AbstractSessionHandler`, which implements the logic to remove/delete cookies when needed + deal with `validateId()`+`updateTimestamp()`
- `SessionHandlerProxy`, which turns any handler into an instance of `AbstractSessionHandler`, especially useful and require for the native session handler, which doesn't implement the special interface.

The rest are related tweaks.

TerjeBr commented:
Have you tested this with a full symfony stack?

When a typical symfony SessionBag is empty it is not the empty string, it contains some metadata.
A new session that is started with no data will contain an array with these 3 keys: `_sf2_attributes`, `_sf2_flashes` and `_sf2_meta`

Because of this, this PR will not fix #6388

stof commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php:
```diff
@@ -517,7 +542,9 @@ private function doRead($sessionId)
 
-            if (self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+            if (!ini_get('session.use_strict_mode') && self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+                // In strict mode, session fixation is not possible: new sessions alway start with a unique
```
typo: always

nicolas-grekas replied:
fixed

stof commented on src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/AbstractSessionHandlerTest.php:
```diff
@@ -0,0 +1,61 @@
+            1 => array('file', '/dev/null', 'w'),
+            2 => array('file', '/dev/null', 'w'),
+        );
+        if (!self::$server = @proc_open('exec php -S localhost:8053', $spec, $pipes, __DIR__.'/Fixtures')) {
```
`exec` won't work on windows

nicolas-grekas replied:
Correct, on purpose: signaling the `php -S` process to kill it is complex, so this is skipped on Windows as I don't want to import the Process component just for this.

nicolas-grekas commented:
@TerjeBr thanks for the hint, that's now fixed (but not tested yet).
That allowed me to spot that the BC layer in place right now is broken. We didn't spot it because of #23003 (PhpUnit-Bridge can't deal with `@runTestsInSeparateProcesses`).
Fixed also.

TerjeBr commented on src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php:
```diff
@@ -217,23 +217,40 @@ public function regenerate($destroy = false, $lifetime = null)
+        $bags = array_merge($this->bags, array($this->metadataBag));
+
+        foreach ($bags as $bag) {
+            if (empty($_SESSION[$key = $bag->getStorageKey()])) {
```
This is a good try, but will not work because the metadataBag is never empty. Once the session is started, the metadataBag is created ans filled with some metedata that always will be there as long as the session lives.

Another concern I have is that this soluton is less general, and every new type of session storage has to be changed to make sure it stores nothing in the session when the bags are empty. This certainly will be a bigger BC break for those that write their own sessionStorage class, than just adding a "wasStarted" or "isActive" method.

nicolas-grekas replied:
Fixed, thanks. The approach in this PR fixes more issues with sessions, and works fine when `$_SESSION` is used directly, so it definitely the way to go to me.

TerjeBr commented on src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php:
```diff
@@ -217,23 +217,41 @@ public function regenerate($destroy = false, $lifetime = null)
+        }
+        if (array($key = $this->metadataBag->getStorageKey()) === array_keys($_SESSION)) {
+            unset($_SESSION[$key]);
+        }
```
If the session is empty at this point, it will be an empty array, right? Will that serialize to an empty string? Or will it be something like `a:0:{}`?

I am a little unfamiliar with how your tests work, could you please be so kind to point out for me the file and line where you test this?

nicolas-grekas replied:
Correct: sessions serialization is special and doesn't use "serialiaze()" but a variant of it. And the empty session array is represented by the empty string (that's why session handlers return the empty string when they have no data for a specific session id.)

TerjeBr replied:
Just out of curiosity, will `null` also serialize to an empty string?

You did not tell me about the test, I assume the full use of a nativeSessionStorage bag is covered in the tests?

nicolas-grekas replied:
"null" is not a valid value for `$_SESSION` so I don't know how PHP will behave if you such such a thing?

nicolas-grekas replied:
Oh, about the tests, there is none right now, I'm on it :)

TerjeBr replied:
I just read up a bit on this, have a look at http://php.net/manual/en/function.session-decode.php and http://php.net/manual/en/session.configuration.php#ini.session.serialize-handler

It seems like this will only be working when  `session.serialize_handler` is set to the default `php` one?

nicolas-grekas replied:
I don't know how the other serialize handler serialize the empty array. Let's change the check and use `empty($_SESSION)` instead.

TerjeBr replied:
Since you need to do that in the `write` function of the session handler, you may want to use this https://packagist.org/packages/wikimedia/php-session-serializer

TerjeBr replied:
Ok, scrap that, you have also things like [session.serialize_handler=igbinary](https://github.com/igbinary/igbinary) so the wikimedia/php-session-serializer would not cover all use cases.

nicolas-grekas replied:
I opened https://github.com/igbinary/igbinary/issues/146 about igbinary, and meanwhile it is now handled in the code. All other serialization handler store the empty session as the empty string.

nicolas-grekas replied:
Now tested, see https://github.com/symfony/symfony/pull/24523/files#diff-df2c28eb378a2dc581bb6b53fff80661

TerjeBr replied:
I do not see where you have tested with all the different session handlers.

And in which test do you test that it does not start a session if just just read from a session bag?

nicolas-grekas replied:
The existing session handlers are already tested, there's no need to test them again. See fixture I linked.

nicolas-grekas replied:
storage.php

TerjeBr commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php:
```diff
@@ -0,0 +1,156 @@
+     */
+    public function write($sessionId, $data)
+    {
+        if ('' === $data) {
```
If I am not missataken, this is where the magic happens that prevents the session from starting if you only read from it and it has not been written to before. Right?

It only tests to see if the data is an empty string. If the data is `null` or an empty array it will not be triggered. Is $data here the same as $_SESSION when you write the session, or is it a serialized verson of it?

TerjeBr replied:
Just another thought: It might be a good thing to give the authors of sessonStorage classes more leeway in what counts as an empty session.

We should also have more documentation of these new things you need to consider when you write a new sessionStorage class, (but I guess that belongs in a separate documentation ticket.)

nicolas-grekas replied:
It's always a string, pre-serialized by PHP, that's the contract and behavior of SessionHandlerInterface.
This catches the empty-session case, either no-writes, or all-deleted yes.

TerjeBr replied:
Ok, thank you for answering my concern about this.

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/PdoSessionHandler.php:
```diff
@@ -517,7 +542,9 @@ private function doRead($sessionId)
-            if (self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+            if (!ini_get('session.use_strict_mode') && self::LOCK_TRANSACTIONAL === $this->lockMode && 'sqlite' !== $this->driver) {
+                // In strict mode, session fixation is not possible: new sessions always start with a unique
+                // random id, so that concurrency is not possible and this code path can be skipped.
```
I don't agree. If the session just expired or was deleted manually, there can still be race conditions.

nicolas-grekas replied:
Can you describe the scenario?
To me, when the session expired, or when the session was deleted in another place,
*in strict mode only*, php will generate a new id, so that there is no race possible.

Tobion replied:
I wonder how php knows if the session is new? so when you do session_start() with an old id which expired, the save handler is executed which returns an empty string. And then php finishes the save handler by calling close() etc and then generates a new session id? Does it call the save handler open, read etc again?

nicolas-grekas replied:
Thanks to the new interface, PHP calls "validateId", which return false if the handler's "read" returns the empty string.
Then, PHP calls "read()" directly, but now with a new random id.

Tobion replied:
This is important also for the advisory locks which might otherwise not be released at all if it just calls save() with a new id.

nicolas-grekas replied:
Anything to change in your opinion? The advisory lock looks also good to me as is.

Tobion replied:
> Thanks to the new interface, PHP calls "validateId", which returns false if the handler's "read" returns the empty string. Then, PHP calls "read()" directly, but now with a new random id.

That does not make sense to me. When I start a new session without having a cookie, PHP generates a new session id which it then calls `validateId` with. As we have no data, it returns false. So php generates a new session id again, and calls `read` which again return no data. So to start a new session, it tries to read two different session ids twice?

nicolas-grekas replied:
Validation happens only if there is a cookie of course.

Tobion replied:
That does not seem to be true. The storage.expected tests shows what I was questioning

```
validateId
read
doRead: 
read
doRead: 
```

and there does not seem to be a cookie set before.

nicolas-grekas replied:
The issue with going with only a proxy is that it will prevent skipping the lock in PdoSessionHandler, because we couldn't know if we're having the required logic or not.

nicolas-grekas replied:
Oh, and implementating updateTimestamp per-handler would have no contract to be based on (and adding one would be worse.)

Tobion replied:
I think the validation is only about checking if the session id has invalid characters in it according to the patch for strict sessions https://gist.github.com/yohgaki/1379668#file-php-trunk-strict-session-patch-L55

Thus I don't think validation should actually read the data.

Tobion replied:
What I wonder is, if php implements a default validation of the id, e.g. by checking the length of the session id against `session.sid_length` php ini. Something like that seems more the correct approach to me.

Tobion replied:
Reading https://wiki.php.net/rfc/strict_sessions

> This patch adds validate_id() function to ps_modules (session save handler) that can check if the session ID is already initialized one or not. Bundled ps_modules are now have validate function and will never accept uninitialized session ID

it suggests validate id is really about checking whether the session is initialized. Then the current solution is fine. I'd suggest to prevent new session from being read twice, to do

```
if ($prefetchId === $sessionId || '' === $prefetchData) {
    return $prefetchData;
}
```

Tobion replied:
Back to the original question. With the validateId and strict session in place, race conditions on new session ids cannot happen indeed. So it's fine to skip inserting fake rows.

nicolas-grekas replied:
> if ($prefetchId === $sessionId || '' === $prefetchData) {

good idea, done

Tobion commented on src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/Fixtures/empty_destroys.expected:
```diff
@@ -0,0 +1,17 @@
+Array
+(
+    [0] => Content-Type: text/plain; charset=utf-8
+    [1] => Cache-Control: private, max-age=10800
```
http://php.net/manual/en/function.session-cache-limiter.php says something about `pre-check`. So php documentation is wrong about that? pre-check doesn't seem to be a standard anyway.

nicolas-grekas replied:
Looks like so yes...

nicolas-grekas commented:
PR green and ready :)

nicolas-grekas commented:
(fixes current failures on 3.4)

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php:
```diff
@@ -0,0 +1,160 @@
+    public function write($sessionId, $data)
+    {
+        if (null === $this->igbinaryEmptyData) {
+            $this->igbinaryEmptyData = \function_exists('igbinary_serialize') ? igbinary_serialize(array()) : '';
```
you could add a link to your igbinary bug report here for clarity

nicolas-grekas replied:
done

TerjeBr commented:
@nicolas-grekas wrote:
> I opened igbinary/igbinary#146 about igbinary, and meanwhile it is now handled in the code. All other serialization handlers store the empty session as the empty string.

What about the `php_serialize` handler? From http://php.net/manual/en/session.configuration.php#ini.session.serialize-handler:
>php_serialize uses plain serialize/unserialize function internally

`serialize(array())` returns `"a:0:{}"` and not an empty string.

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php:
```diff
@@ -0,0 +1,160 @@
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+abstract class AbstractSessionHandler implements \SessionHandlerInterface, \SessionUpdateTimestampHandlerInterface
```
i think this class can be implemented as a proxyhandler as well similar to SessionHandlerWrapper that wraps another SessionHandlerInterface. just check that the wrapped handler  does not implement SessionUpdateTimestampHandlerInterface. or alternatively, in case it implements SessionUpdateTimestampHandlerInterface just proxy those method as well. that might even be better because this class `destroy` logic might be useful even for those handlers.

Tobion replied:
this way, the other session handlers don't need to be changed and they don't need to know about `$this->sessionName`

nicolas-grekas replied:
(re-posting answer on the correct comment)
The issue with going with only a proxy is that it will prevent skipping the lock in PdoSessionHandler, because we couldn't know if we're having the required logic or not.
Oh, and implementating updateTimestamp per-handler would have no contract to be based on (and adding one would be worse.)

Tobion replied:
I don't see the problem with the PdoSessionHandler. Instead of AbstractSessionHandler::doRead and doWrite etc, just call `$this->wrapperSessionHandler->read()`

Tobion replied:
And for updateTimestamp: 
```
if ($this->wrapperSessionHandler instanceof SessionUpdateTimestampHandlerInterface) {
   $this->wrapperSessionHandler ->updateTimestamp();
} else {
    $this-save();
}
```

nicolas-grekas replied:
> And for updateTimestamp:

that implies implementing `SessionUpdateTimestampHandlerInterface` on existing handlers, so would require implementing also "validateId". That's what the abstract is for.

Tobion replied:
validateId can be an empty method as it is not called. SessionUpdateTimestampHandlerInterface is implemented in exsting handlers anyway by extending the abstract. another advantage is that this proxy is usable with any existing session handlers even when they not come from symfony.

nicolas-grekas replied:
I don't see any benefit to removing the abstract. Creating dummy functions is not better, what would this improve?
SessionHandlersWrapper is already usable with any existing handler, so this need is already filled.
What are you seeking for?

Tobion replied:
So it only needs one wrapper. Having both an abstract and a wrapper for no reason just makes the api confusing. Also the naming of these classes should be improved. abstract and wrapper just does not say anything about what they achieve which should be clear from the class name.

nicolas-grekas replied:
> going with only a proxy is that it will prevent skipping the lock in PdoSessionHandler

The issue here is that the handler (pdo in this case) will not be able to make the assumption that session ids are unique. Thus, we would not be able to skip the locking logic in PDO.

I really don't see the point of not having the abstract for the sake of it sorry The wrapper is needed to add the strict behavior to the native proxy (too bad it doesn't have it already), and is useful for decorating existing handlers of course. About the name, suggestions welcomed (naming is hard :) )

nicolas-grekas replied:
> validateId can be an empty method as it is not called

as soon as `SessionUpdateTimestampHandlerInterface` is implemented, it *can* be called, unless we want to provide deficient implementation that must be wrapped. That's why the abstract is needed: to provide a working implement of it.

Tobion replied:
This class needs phpdoc to explain what it achieves.

Tobion commented on src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/Fixtures/storage.expected:
```diff
@@ -0,0 +1,20 @@
+    [0] => bar
+)
+$_SESSION is not empty
+write
```
I thought the lazy_write ini does not call `write` if data didn't change. The data was empty at the start of the session and it is empty at the end (because the flash got added and removed again by reading it). So why does it call write anyway?

nicolas-grekas replied:
lazy write calls updateTimestamp instead of write, but only when the session is not empty.
When it's empty, it calls write.

nicolas-grekas commented:
> What about the php_serialize handler? 

I wondered the same, so I tried: the empty session is serialized as the empty string.
That's consistent with the source code:
https://github.com/php/php-src/blob/master/ext/session/session.c#L843

TerjeBr commented on src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/Fixtures/storage.php:
```diff
@@ -0,0 +1,24 @@
+$storage = new NativeSessionStorage();
+$storage->setSaveHandler(new TestSessionHandler());
+$flash = new FlashBag();
+$storage->registerBag($flash);
```
Here you test NativeSessionStorage with one flash bag. That is good.

But I miss a test where you have the full symfony framework, where the standard bags are automatically added, including a metaData bag.

Will this test catch it if the metaData bag is not removed when all the other bags are empty?

Tobion replied:
@TerjeBr the metadata bag is automatically added by the constructor of the nativesessionstorage. there is no difference to full symfony. see nativesessionstorage::save that deletes empty bags.

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php:
```diff
@@ -377,37 +395,29 @@ public function setOptions(array $options)
-        if (!$saveHandler instanceof AbstractProxy && $saveHandler instanceof \SessionHandlerInterface) {
+        if (null === $saveHandler) {
+            $saveHandler = new SessionHandlerWrapper(new \SessionHandler());
+        } elseif (!$saveHandler instanceof \SessionHandlerInterface) {
```
doesn't this break when session handler is `AbstractProxy`. AbstractProxy does not implement \SessionHandlerInterface but SessionHandlerProxy requires a \SessionHandlerInterface

nicolas-grekas replied:
fixed

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php:
```diff
@@ -0,0 +1,163 @@
+     *
+     * @return string
+     */
+    public function getId()
```
getId and isActive are only implemented for bc with AbstractProxy right? I guess those method should be marked as deprecated

sroze commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 3.4.0
 -----
 
+ * Session `use_strict_mode` is now enabled by default and the corresponding option has been deprecated,
```
There is no consequence at all with existing sessions or any BC break?

nicolas-grekas replied:
No technical consequences, only behaviorals, if session fixation is a feature for you (meaning you're putting your users at risk.) Can you think of any BC issue?

phpduck replied:
@nicolas-grekas I can think of a BC: SSO!

nicolas-grekas replied:
Did you experience it? because "session fixation" is not a feature we want to support IMHO.

phpduck replied:
Yes, e.g. with this https://github.com/legalthings/sso - There is nothing wrong with keeping session fixation enabled for SSO, PHP's default setting for strict_mode is 0 and there are plenty of reasons for that. Could you explain why you suddenly think session fixation is unsecure after all these years?

nicolas-grekas replied:
If you want to open that discussion, please open an issue.

phpduck replied:
Switched to Laravel, all good now.

nicolas-grekas commented:
@Tobion the deprecation layer of `Session\Storage\Proxy\*` classes is completely broken: see `NativeSessionStorage::setId()` which calls `$this->saveHandler->setId($id)`.
I'm going to revert this deprecation in this PR, could re-added later, but fixing is not trivial so not for this PR.

nicolas-grekas commented:
Deprecation of `SessionHandlerProxy` and `AbstractProxy` now reverted. It's too much work for this PR, so should be done is a separate one. For 4.1 is fine.

chalasr commented:
I confirm the broken BC layer, core classes still extend/use them. Deprecating them in 4.1 looks sensible to me.

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MemcachedSessionHandler.php:
```diff
@@ -65,6 +65,8 @@ public function __construct(\Memcached $memcached, array $options = array())
      */
     public function open($savePath, $sessionName)
     {
+        $this->sessionName = $sessionName;
```
that could be implemented in the parent class. and then just call paren::open. then sessionName does not need to be protected as well.

nicolas-grekas replied:
good idea, done

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * {@inheritdoc}
```
i thought we don't add inheritdoc anymore which seems useless.

nicolas-grekas replied:
that was asked for, but rejected AFAIK

Tobion commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -462,7 +462,10 @@ private function addSessionSection(ArrayNodeDefinition $rootNode)
+                            ->setDeprecated('The "%path%.%node%" option is enabled by default and deprecated since Symfony 3.4. It will be always enabled in 4.0.')
+                        ->end()
                         ->scalarNode('save_path')->defaultValue('%kernel.cache_dir%/sessions')->end()
                         ->integerNode('metadata_update_threshold')
```
the info of this option is wrong now

> it will also prevent the session handler to write if the session has not changed

Tobion replied:
also as lazy_write is only available since php 7, users with lower php versions will not have lazy write anymore

nicolas-grekas replied:
Indeed. I just added an emulation of lazy writes in the abstract handler for PHP 5.

Tobion commented on UPGRADE-4.0.md:
```diff
@@ -542,10 +542,11 @@ HttpFoundation
-   `Symfony\Component\HttpFoundation\Session\Storage\Proxy\NativeProxy` and
-   `Symfony\Component\HttpFoundation\Session\Storage\Proxy\SessionHandlerProxy` classes have been removed.
+ * The `Symfony\Component\HttpFoundation\Session\Storage\Handler\WriteCheckSessionHandler` class has been
+   removed. Implement `SessionUpdateTimestampHandlerInterface` or extend `AbstractSessionHandler` instead.
```
those or no alternative for WriteCheckSessionHandler. only session.lazy_write option is.

nicolas-grekas replied:
`session.lazy_write` requires `updateTimestamp()`, thus `SessionUpdateTimestampHandlerInterface`

nicolas-grekas commented:
comments addressed, still green :)

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php:
```diff
@@ -100,8 +100,7 @@ class NativeSessionStorage implements SessionStorageInterface
         $options += array(
-            // disable by default because it's managed by HeaderBag (if used)
-            'cache_limiter' => '',
+            'cache_limiter' => 'private_no_expire',
```
This was implicit before because caches usually don't store responses with "Set-Cookie"; but now that we don't send the cookie, we must tell proxies to not store the page, otherwise eg. flashes will never be shown (as the cached response will be served).
Of course, apps can take over and set another Cache-Control, since HeaderBag applies after the session started.

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/SessionHandlerWrapper.php:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class SessionHandlerWrapper extends AbstractSessionHandler
```
can't we come up with an explicit name like `SessionFixationPreventionAdapter`

nicolas-grekas replied:
Pushed as `StrictSessionHandler`, docblock added on `AbstractSessionHandler`.

chalasr commented on src/Symfony/Component/HttpFoundation/CHANGELOG.md:
```diff
@@ -4,8 +4,8 @@ CHANGELOG
- * deprecated the `NativeSessionHandler` class,
- * deprecated the `AbstractProxy`, `NativeProxy` and `SessionHandlerProxy` classes,
+ * implemented PHP 7.0's `SessionUpdateTimestampHandlerInterface` with a new `AbstractSessionHandler`
+ * deprecated the `WriteCheckSessionHandler`, `NativeSessionHandler` and `NativeProxy` classes
```
I would add an entry for `StrictSessionHandler`, it's useful

chalasr reviewed (approved):
👍

fabpot commented on src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 3.4.0
 -----
 
+ * Session `use_strict_mode` is now enabled by default and the corresponding option has been deprecated,
```
extra comma at the end

fabpot replied:
Looks like there is a missing counterpart of the CHANGELOG in `UPGRADE-4.0.md`

Tobion reviewed (approved):
Nice work @nicolas-grekas. Improving the session system was long overdue and I like how you solved it.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -908,24 +908,17 @@ private function registerSessionConfiguration(array $config, ContainerBuilder $c
-            }
-
-            $container->setAlias('session.handler', $handlerId)->setPrivate(true);
+            $options['lazy_write'] = 1;
```
Why is lazy write only enabled in this case but not for native session handlers? Seems arbitrary as we don't know more about a custom session handler than we do about the native one. So lazy_write should be safe to be enabled all the time.

Tobion commented on src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php:
```diff
@@ -401,11 +408,13 @@ public function setSaveHandler($saveHandler = null)
             $saveHandler = new SessionHandlerProxy($saveHandler);
         } elseif (!$saveHandler instanceof AbstractProxy) {
-            $saveHandler = new SessionHandlerProxy(new \SessionHandler());
+            $saveHandler = new SessionHandlerProxy(new StrictSessionHandler(new \SessionHandler()));
```
It feels strange that we need to make the native session andlers strict. The `files` session handler is already strict I assume. So this is only needed because if you use `\SessionHandler`, you lose that strictness? \SessionHandler does not implement `\SessionUpdateTimestampHandlerInterface`! So if we not not set a save handler at all it would even be better as it would then use the validate id implementation of the native session handler? But currently that is not possible do.

nicolas-grekas replied:
Correct, I don't think we can do better here, id validation works with the wrapper so not a big issue. There is only updateTimestamp which ships no optimization but only writes. Not sure we can do better.

Tobion replied:
If php doesn't change this behavior, we can think about deprecating NativeFilesSessionHandler in SF 4.1. So by default it just sets the session handler without actually overwriting \SessionHandler.

TerjeBr commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php:
```diff
@@ -171,7 +163,34 @@ public function write($sessionId, $data)
-    public function read($sessionId)
+    public function updateTimestamp($sessionId, $data)
+    {
+        $expiry = $this->createDateTime(time() + (int) ini_get('session.gc_maxlifetime'));
```
This is duplicated code from function `doWrite` (line 144). Would it not be better to put this common code in a new protected method? May be something like `createExpiryTime`.

nicolas-grekas replied:
not worth it for one line IMHO

TerjeBr commented on src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php:
```diff
@@ -217,23 +216,38 @@ public function regenerate($destroy = false, $lifetime = null)
+            }
+        }
+        if (array($key = $this->metadataBag->getStorageKey()) === array_keys($_SESSION)) {
+            unset($_SESSION[$key]);
```
May be you should use `$session` here instead of `$_SESSION`? That way the metadata is not destroyed until the `finally` clause at line 241

nicolas-grekas commented:
comments addressed

fabpot commented:
Thank you @nicolas-grekas.

Tobion commented:
@nicolas-grekas what I proposed in #12325 is to not start the session at all until data is set when there is no session cookie in the request. I don't think that it implemented yet. This would avoid generating a session id, starting the session handler etc when you just read session data to then realize nothing is there.

nicolas-grekas commented:
@Tobion from the HTTP pov, the observed behavior is exactly the same. In fact, starting the session has the benefit of sending the appropriate Cache-Control header. If HTTP cacheability is improved by the current change, it means we may not have to care anymore about whether the session is really started or not.

Tobion commented:
If I just want to check if the user is logged in to then redirect or whatever, I don't need to start the session if there is no session cookie. And a cache control header might not be what I want.
